[AP][IP] Split Initial Placement Into Stages

AlexandreSinger opened the pull request:
In the AP flow, the initial placement stage will try to create a minimum displacement placement based on the flat placement and the clustering. It does this by iteratively trying to place clusters farther and farther from where they want to be placed (according to the flat placement).

I found that this flow was having issues with macros that contained multiple clusters. If these blocks are placed too late, they may not find a hole big enough for them to fit into. To resolve this issue, split the initial placer into stages such that all fixed blocks are placed first (this blocks must be placed first to ensure other blocks don't take their spots), then all large macros are placed, then all remaining clusters are placed.

I expected to make this a backup flow, which only turned on if the first flow failed; however, I found that this actually outperformed my original implementation. I think that placing these larger macros early (i.e. placing them closer to their flat placement position) allows more blocks to be placed closer to where they want to be placed.

For most circuits this had a very minor improvement, but for other circuits (such as MCML from the VTR benchmarks) they improved massively.

Results on VTR with no fixed blocks and not timing driven:

  | post_fl_hpwl | post_dp_hpwl | total_wirelength | post_fl_avg_disp | post_fl_max_disp | total_runtime
-- | -- | -- | -- | -- | -- | --
arm_core.v | 1.00 | 0.98 | 0.99 | 1.00 | 1.00 | 1.01
bgm.v | 1.00 | 0.99 | 1.00 | 0.99 | 0.96 | 1.01
mkDelayWorker32B.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.01
stereovision0.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.00
stereovision1.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.01
stereovision2.v | 0.99 | 1.02 | 1.01 | 1.00 | 1.00 | 1.01
LU8PEEng.v | 0.98 | 1.00 | 0.99 | 0.99 | 0.98 | 0.99
LU32PEEng.v | 0.97 | 0.92 | 0.94 | 0.96 | 1.01 | 0.96
mcml.v | 0.58 | 0.99 | 1.00 | 0.63 | 0.53 | 0.92

As you can see, most circuits remained mostly the same (all improvements), but the circuits with many large macros improved. with MCML getting a huge improvement. MCML was the one that was causing issues, so it makes sense it would improve the most.

vaughnbetz commented on vpr/src/place/initial_placement.cpp:
```diff
@@ -1844,13 +1861,126 @@ static inline void place_all_blocks_ap(enum e_pad_loc_type pad_loc_type,
+    }
+
+    if (constrained_clusters.size() > 0) {
+        VTR_LOG("Placing constrained clusters...\n");
```
- You could blend macro size and region size to get a better overall sort order. E.g. macro.size + 1/blocks_in_region could be the overall sort criterion.

Advantages:    Bigger macros are harder. Smaller regions are harder. This blends both.

vaughnbetz reviewed (approved):
Ok with me. I have a suggestion for a further improvement but it would take more code.

AlexandreSinger commented:
Titan results for the initial placement changes. No fixed blocks, not timing driven:

Circuit | post_fl_hpwl | post_dp_hpwl | total_wirelength | post_fl_avg_disp | post_fl_max_disp | total_runtime
-- | -- | -- | -- | -- | -- | --
LU230_stratixiv_arch_timing.blif | 1.00 | 1.01 | 1.01 | 1.00 | 1.00 | 0.98
LU_Network_stratixiv_arch_timing.blif | 1.00 | 0.98 | 0.99 | 1.00 | 1.02 | 0.98
SLAM_spheric_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.94
bitcoin_miner_stratixiv_arch_timing.blif | 0.99 | 1.00 | 1.00 | 0.99 | 1.00 | 1.05
bitonic_mesh_stratixiv_arch_timing.blif | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 1.01
cholesky_bdti_stratixiv_arch_timing.blif | 1.00 | 1.01 | 1.01 | 1.00 | 1.00 | 0.99
cholesky_mc_stratixiv_arch_timing.blif | 1.00 | 0.97 | 0.98 | 0.99 | 1.00 | 0.86
dart_stratixiv_arch_timing.blif | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 0.96
denoise_stratixiv_arch_timing.blif | 1.00 | 1.02 | 1.01 | 1.00 | 1.00 | 0.96
des90_stratixiv_arch_timing.blif | 1.00 | 1.02 | 1.02 | 1.00 | 1.00 | 0.91
directrf_stratixiv_arch_timing.blif | 0.95 | 0.97 | 0.97 | 0.84 | 1.00 | 1.07
gsm_switch_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.97
mes_noc_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.97
minres_stratixiv_arch_timing.blif | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 0.96
neuron_stratixiv_arch_timing.blif | 1.00 | 1.03 | 1.03 | 1.00 | 1.00 | 0.85
openCV_stratixiv_arch_timing.blif | 1.00 | 1.02 | 1.01 | 1.00 | 1.00 | 0.93
segmentation_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.99
sparcT1_chip2_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.98
sparcT1_core_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.01 | 1.00 | 1.00 | 0.92
sparcT2_core_stratixiv_arch_timing.blif | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 0.97
stap_qrd_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 0.96 | 1.00 | 0.92
stereo_vision_stratixiv_arch_timing.blif | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.81
GEOMEAN | 1.00 | 1.00 | 1.00 | 0.99 | 1.00 | 0.95

Basically a perfect wash over all circuits, with minor improvements in some of the largest circuits like DirectRF and bitcoin_miner. I think the run time improvement may mostly be coming from machine load, but I am not 100% sure. Either way these changes did not harm anything and it makes the initial placer more robust.

amin1377 commented:
@AlexandreSinger: Have you also tried this on Koios? I’m asking because the number of chains in the Koios benchmarks is higher than in the other benchmarks you’ve tested, and it would be interesting to see how this impacts the results.

AlexandreSinger commented:
To whom it may concern (i.e. @amin1377 ), these are the results for Koios (wirelength driven, with fixed blocks):

  | post_fl_hpwl | post_dp_hpwl | total_wirelength | post_fl_avg_disp | post_fl_max_disp | total_runtime
-- | -- | -- | -- | -- | -- | --
attention_layer.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.01
bnn.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.06
bwave_like.fixed.large.v | 0.98 | 1.01 | 1.01 | 1.00 | 0.94 | 1.10
bwave_like.fixed.small.v | 1.00 | 0.99 | 1.00 | 1.00 | 1.00 | 0.96
bwave_like.float.large.v | 0.98 | 0.99 | 1.00 | 1.00 | 1.00 | 1.02
bwave_like.float.small.v | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 1.04
clstm_like.large.v | 1.00 | 0.98 | 0.97 | 1.00 | 1.00 | 0.97
clstm_like.medium.v | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 0.95
clstm_like.small.v | 1.00 | 0.99 | 1.00 | 1.00 | 1.00 | 0.89
conv_layer.v | 1.00 | 0.97 | 0.97 | 1.00 | 1.00 | 0.95
conv_layer_hls.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.04
dla_like.large.v | 0.96 | 0.98 | 0.99 | 1.01 | 0.83 | 1.00
dla_like.medium.v | 0.95 | 1.00 | 1.00 | 1.03 | 1.12 | 1.03
dla_like.small.v | 0.95 | 0.99 | 1.00 | 1.03 | 1.00 | 1.02
dnnweaver.v | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 1.05
eltwise_layer.v | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 1.17
gemm_layer.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.03
lenet.v | 1.04 | 1.00 | 1.00 | 0.99 | 1.00 | 1.01
lstm.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.06
reduction_layer.v | 1.00 | 0.99 | 0.99 | 1.00 | 1.00 | 1.01
robot_rl.v | 0.99 | 0.96 | 0.97 | 0.99 | 1.00 | 0.98
softmax.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 0.96
spmv.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.01
tdarknet_like.large.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.01
tdarknet_like.small.v | 1.00 | 1.00 | 1.00 | 1.00 | 1.00 | 1.03
tpu_like.large.os.v | 0.99 | 1.01 | 1.00 | 1.00 | 1.00 | 0.96
tpu_like.large.ws.v | 1.02 | 1.11 | 1.09 | 1.00 | 1.00 | 1.04
tpu_like.small.os.v | - | - | - | - | - | -
tpu_like.small.ws.v | - | - | - | - | - | -
GEOMEAN | 0.99 | 1.00 | 1.00 | 1.00 | 1.00 | 1.01

Note that 2 circuits are failing. They are failing at initial placement. I think they are having some difficulties placing the solution from APPack; it could be that the device was sized very small.

Overall practically a wash, with minor improvements in post-FL HPWL. Still good since this resolves some issues with the Initial Placer.

amin1377 commented:
Interesting! Thanks, Alex!
